client/src/librarySymbolManager.ts

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as LC from "vscode-languageclient/node";
+import * as path from "path";
+import * as vscode from "vscode";
+
+import { LibraryJson, PowerQueryApi } from "./vscode-powerquery.api";
+import { Convert } from "./libraryExportJSON";
+
+const ErrorMessagePrefix: string = "Error processing symbol directory path. Please update your configuration.";
+
+const SymbolFileExtension: string = ".json";
+const SymbolFileEncoding: string = "utf-8";
+
+export type MinimalClientTrace = Pick<LC.BaseLanguageClient, "debug" | "info" | "warn" | "error">;
+export type MinimalFileSystem = Pick<vscode.FileSystem, "readDirectory" | "readFile" | "stat">;
+
+export class LibrarySymbolManager {
+    private readonly registeredSymbolModules: string[] = [];
+    private readonly fs: vscode.FileSystem;
+
+    constructor(
+        private librarySymbolClient: PowerQueryApi,
+        private clientTrace?: MinimalClientTrace,
+        fs?: vscode.FileSystem,
+    ) {
+        this.fs = fs ?? vscode.workspace.fs;
+    }
+
+    public async refreshSymbolDirectories(directories?: string[]): Promise<readonly string[]> {
+        await this.clearAllRegisteredSymbolModules();
+
+        if (!directories || directories.length === 0) {
+            return [];
+        }
+
+        // Fetch the full list of files to process.
+        const fileDiscovery: Promise<vscode.Uri[]>[] = [];
+
+        const directoryUris: vscode.Uri[] = directories.map((d: string) => {
+            const normalized: string = path.normalize(d);
+
+            if (d !== normalized) {
+                this.clientTrace?.info(`Normalized symbol file path '${d}' => '${normalized}'`);
+            }
+
+            return vscode.Uri.file(normalized);
+        });
+
+        // TODO: Remove duplicate directories?
+
+        directoryUris.forEach((d: vscode.Uri) => {
+            fileDiscovery.push(this.getSymbolFilesFromDirectory(d));
+        });
+
+        // TODO: check for duplicate module file names and only take the last one.
+        // This would allow a connector developer to override a symbol library generated
+        // with an older version of their connector.
+        const symbolFileActions: Promise<[vscode.Uri, LibraryJson | undefined]>[] = [];
+        const files: vscode.Uri[] = (await Promise.all(fileDiscovery)).flat();
+
+        files.forEach((fileUri: vscode.Uri) => {
```

Comment:
Use "for of" over "forEach"

Reply:
Actually, in this case we don't even need a for loop:

```
const files: vscode.Uri[] = (await Promise.all(fileDiscovery)).flat();
const allSymbolFiles = await ArrayUtils,mapAsync(files, async file => ...)
```

Reply:
I tried this but combined with the change to make `processSymbolFile` return `[vscode.Uri, LibraryJson] | undefined`, it makes the code awkward because we need to filter out the `undefined` results.